File: frontend/src/types/financial.ts
export enum TransactionType {
  TIP = 'tip',
  PPV = 'ppv',
  SUBSCRIPTION = 'subscription',
  MESSAGE = 'message',
  REFERRAL = 'referral',
  ADJUSTMENT = 'adjustment',
  CHARGEBACK = 'chargeback',
  REFUND = 'refund',
}

export enum TransactionStatus {
  PENDING = 'pending',
  COMPLETED = 'completed',
  FAILED = 'failed',
  REFUNDED = 'refunded',
  DISPUTED = 'disputed',
}

export enum PayoutStatus {
  PENDING = 'pending',
  PROCESSING = 'processing',
  COMPLETED = 'completed',
  FAILED = 'failed',
  CANCELLED = 'cancelled',
}

export enum PaymentMethodType {
  BANK_TRANSFER = 'bank_transfer',
  PAYPAL = 'paypal',
  CRYPTO = 'crypto',
  CHECK = 'check',
  WIRE = 'wire',
}

export interface Transaction {
  id: number;
  agency_id: number;
  model_id: number;
  type: TransactionType;
  status: TransactionStatus;
  platform_transaction_id?: string;
  conversation_id?: number;
  message_id?: number;
  gross_amount: number;
  platform_fee: number;
  agency_commission: number;
  net_amount: number;
  currency: string;
  fan_id: string;
  fan_username: string;
  transaction_date: string;
  processed_at?: string;
  description?: string;
  metadata?: Record<string, any>;
  created_at: string;
  updated_at: string;
}

export interface Earning {
  id: number;
  model_id: number;
  transaction_id: number;
  period_year: number;
  period_month: number;
  period_week: number;
  gross_amount: number;
  commission_amount: number;
  net_amount: number;
  is_paid: boolean;
  payout_id?: number;
  paid_date?: string;
  created_at: string;
  updated_at: string;
}

export interface Payout {
  id: number;
  agency_id: number;
  model_id: number;
  payout_number: string;
  status: PayoutStatus;
  period_start: string;
  period_end: string;
  gross_earnings: number;
  commission_amount: number;
  adjustments: number;
  net_amount: number;
  currency: string;
  payment_method: PaymentMethodType;
  payment_details: Record<string, any>;
  scheduled_date: string;
  processed_date?: string;
  completed_date?: string;
  transaction_reference?: string;
  notes?: string;
  created_at: string;
  updated_at: string;
}

export interface BankDetails {
  account_holder: string;
  account_number: string;
  routing_number?: string;
  bank_name: string;
  swift_code?: string;
  iban?: string;
}

export interface PaymentMethod {
  id: number;
  model_id?: number;
  agency_id?: number;
  method_type: PaymentMethodType;
  is_primary: boolean;
  is_active: boolean;
  nickname?: string;
  
  // Bank transfer
  bank_name?: string;
  account_holder_name?: string;
  account_number?: string; // Masked in responses
  routing_number?: string; // Masked in responses
  swift_code?: string;
  iban?: string; // Masked in responses
  
  // PayPal
  paypal_email?: string;
  
  // Crypto
  crypto_currency?: string;
  crypto_address?: string; // Partially masked in responses
  crypto_network?: string;
  
  // Wire
  wire_instructions?: Record<string, any>;
  
  // Verification
  is_verified: boolean;
  verified_at?: string;
  verification_notes?: string;
  
  // Settings
  minimum_payout: number;
  processing_days: number;
  
  // Metadata
  last_used_at?: string;
  usage_count: number;
  created_at: string;
  updated_at: string;
  
  // Computed property for display
  display_name?: string;
}

export interface PayPalDetails {
  email: string;
}

export interface CryptoWalletDetails {
  currency: string;
  address: string;
  network?: string;
}

export interface Revenue {
  period: 'daily' | 'weekly' | 'monthly' | 'yearly';
  start_date: string;
  end_date: string;
  gross_revenue: number;
  net_revenue: number;
  commission: number;
  currency: string;
  breakdown: RevenueBreakdown[];
}

export interface RevenueBreakdown {
  source: 'subscriptions' | 'tips' | 'messages' | 'content_sales' | 'other';
  amount: number;
  count: number;
  percentage: number;
}

export interface Commission {
  rate: number;
  tiers?: CommissionTier[];
  model_id?: string;
  agency_id?: string;
  effective_date: string;
}

export interface CommissionTier {
  min_revenue: number;
  max_revenue?: number;
  rate: number;
}

export interface Invoice {
  id: string;
  invoice_number: string;
  date: string;
  due_date: string;
  status: 'draft' | 'sent' | 'paid' | 'overdue' | 'cancelled';
  total: number;
  subtotal: number;
  tax: number;
  currency: string;
  items: InvoiceItem[];
  from: InvoiceParty;
  to: InvoiceParty;
  payment_terms?: string;
  notes?: string;
}

export interface InvoiceItem {
  description: string;
  quantity: number;
  unit_price: number;
  total: number;
}

export interface InvoiceParty {
  name: string;
  address?: string;
  email?: string;
  phone?: string;
  tax_id?: string;
}

export interface FinancialReportData {
  earnings?: Record<string, number>;
  deductions?: Record<string, number>;
  taxes?: Record<string, number>;
  commissions?: Record<string, number>;
  summary?: Record<string, string | number>;
}

export interface FinancialReport {
  id: string;
  type: 'earnings_statement' | 'tax_document' | 'commission_report';
  period_start: string;
  period_end: string;
  generated_at: string;
  file_url?: string;
  data: FinancialReportData;
}

export interface FinancialSummary {
  total_revenue: number;
  total_payouts: number;
  pending_payouts: number;
  available_balance: number;
  currency: string;
  last_payout?: Payout;
  next_payout_date?: string;
}

File: backend/models/financial.py
from sqlalchemy import Column, String, Boolean, Integer, ForeignKey, Numeric, JSON, Enum as SQLEnum, Index
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
import enum
from core.database import Base
from models.base import BaseModel


class TransactionType(str, enum.Enum):
    """Transaction type enumeration."""
    TIP = "tip"
    PPV = "ppv"  # Pay-per-view
    SUBSCRIPTION = "subscription"
    MESSAGE = "message"
    REFERRAL = "referral"
    ADJUSTMENT = "adjustment"
    CHARGEBACK = "chargeback"
    REFUND = "refund"


class TransactionStatus(str, enum.Enum):
    """Transaction status enumeration."""
    PENDING = "pending"
    COMPLETED = "completed"
    FAILED = "failed"
    REFUNDED = "refunded"
    DISPUTED = "disputed"


class PayoutStatus(str, enum.Enum):
    """Payout status enumeration."""
    PENDING = "pending"
    PROCESSING = "processing"
    COMPLETED = "completed"
    FAILED = "failed"
    CANCELLED = "cancelled"


class PaymentMethod(str, enum.Enum):
    """Payment method enumeration."""
    BANK_TRANSFER = "bank_transfer"
    PAYPAL = "paypal"
    CRYPTO = "crypto"
    CHECK = "check"
    WIRE = "wire"


class Transaction(BaseModel):
    """Financial transaction record."""
    __tablename__ = "transactions"
    
    # Relationships
    agency_id = Column(Integer, ForeignKey("agencies.id", ondelete="CASCADE"), nullable=False)
    model_id = Column(Integer, ForeignKey("models.id", ondelete="CASCADE"), nullable=False)
    
    # Transaction details
    type = Column(SQLEnum(TransactionType), nullable=False, index=True)
    status = Column(SQLEnum(TransactionStatus), default=TransactionStatus.PENDING, nullable=False)
    
    # External references
    platform_transaction_id = Column(String(255), unique=True, nullable=True)
    conversation_id = Column(Integer, ForeignKey("conversations.id", ondelete="SET NULL"), nullable=True)
    message_id = Column(Integer, ForeignKey("messages.id", ondelete="SET NULL"), nullable=True)
    
    # Financial amounts
    gross_amount = Column(Numeric(12, 2), nullable=False)
    platform_fee = Column(Numeric(12, 2), default=0.00, nullable=False)
    agency_commission = Column(Numeric(12, 2), default=0.00, nullable=False)
    net_amount = Column(Numeric(12, 2), nullable=False)  # Model's earning
    currency = Column(String(3), default="USD", nullable=False)
    
    # Fan information
    fan_id = Column(String(255), nullable=False, index=True)
    fan_username = Column(String(255), nullable=False)
    
    # Timing
    transaction_date = Column(String(30), nullable=False)  # When transaction occurred
    processed_at = Column(String(30), nullable=True)  # When we processed it
    
    # Additional data
    description = Column(String(500), nullable=True)
    transaction_metadata = Column("metadata", JSON, default=dict, nullable=False)
    
    # Relationships
    agency = relationship("Agency", back_populates="transactions")
    model = relationship("Model", back_populates="transactions")
    earning = relationship("Earning", back_populates="transaction", uselist=False)
    
    # Indexes
    __table_args__ = (
        Index('idx_transactions_date', 'transaction_date'),
        Index('idx_transactions_model_date', 'model_id', 'transaction_date'),
    )
    
    def __repr__(self):
        return f"<Transaction {self.type} - ${self.gross_amount} - {self.status}>"


class Earning(BaseModel):
    """Model earnings tracking."""
    __tablename__ = "earnings"
    
    model_id = Column(Integer, ForeignKey("models.id", ondelete="CASCADE"), nullable=False)
    transaction_id = Column(Integer, ForeignKey("transactions.id", ondelete="CASCADE"), unique=True, nullable=False)
    
    # Period tracking
    period_year = Column(Integer, nullable=False)
    period_month = Column(Integer, nullable=False)
    period_week = Column(Integer, nullable=False)
    
    # Amounts
    gross_amount = Column(Numeric(12, 2), nullable=False)
    commission_amount = Column(Numeric(12, 2), nullable=False)
    net_amount = Column(Numeric(12, 2), nullable=False)
    
    # Status
    is_paid = Column(Boolean, default=False, nullable=False)
    payout_id = Column(Integer, ForeignKey("payouts.id", ondelete="SET NULL"), nullable=True)
    
    # Relationships
    model = relationship("Model", back_populates="earnings")
    transaction = relationship("Transaction", back_populates="earning")
    payout = relationship("Payout", back_populates="earnings")
    
    # Indexes
    __table_args__ = (
        Index('idx_earnings_period', 'model_id', 'period_year', 'period_month'),
        Index('idx_earnings_unpaid', 'model_id', 'is_paid'),
    )
    
    def __repr__(self):
        return f"<Earning Model:{self.model_id} - ${self.net_amount} - {'Paid' if self.is_paid else 'Unpaid'}>"


class Payout(BaseModel):
    """Payout to models."""
    __tablename__ = "payouts"
    
    agency_id = Column(Integer, ForeignKey("agencies.id", ondelete="CASCADE"), nullable=False)
    model_id = Column(Integer, ForeignKey("models.id", ondelete="CASCADE"), nullable=False)
    
    # Payout details
    payout_number = Column(String(50), unique=True, nullable=False)
    status = Column(SQLEnum(PayoutStatus), default=PayoutStatus.PENDING, nullable=False)
    
    # Period
    period_start = Column(String(30), nullable=False)
    period_end = Column(String(30), nullable=False)
    
    # Amounts
    gross_earnings = Column(Numeric(12, 2), nullable=False)
    commission_amount = Column(Numeric(12, 2), nullable=False)
    adjustments = Column(Numeric(12, 2), default=0.00, nullable=False)
    net_amount = Column(Numeric(12, 2), nullable=False)
    currency = Column(String(3), default="USD", nullable=False)
    
    # Payment method
    payment_method = Column(SQLEnum(PaymentMethod), nullable=False)
    payment_details = Column(JSON, default=dict, nullable=False)  # Encrypted in production
    
    # Processing
    scheduled_date = Column(String(30), nullable=False)
    processed_date = Column(String(30), nullable=True)
    completed_date = Column(String(30), nullable=True)
    
    # References
    transaction_reference = Column(String(255), nullable=True)  # Bank/payment processor reference
    notes = Column(String(1000), nullable=True)
    
    # Relationships
    agency = relationship("Agency", back_populates="payouts")
    model = relationship("Model", back_populates="payouts")
    earnings = relationship("Earning", back_populates="payout")
    
    def __repr__(self):
        return f"<Payout {self.payout_number} - ${self.net_amount} - {self.status}>"
    
    @property
    def earnings_count(self):
        """Get count of earnings in this payout."""
        return len(self.earnings)


class Invoice(BaseModel):
    """Invoice for agency billing."""
    __tablename__ = "invoices"
    
    agency_id = Column(Integer, ForeignKey("agencies.id", ondelete="CASCADE"), nullable=False)
    
    # Invoice details
    invoice_number = Column(String(50), unique=True, nullable=False)
    invoice_date = Column(String(30), nullable=False)
    due_date = Column(String(30), nullable=False)
    
    # Billing period
    period_start = Column(String(30), nullable=False)
    period_end = Column(String(30), nullable=False)
    
    # Amounts
    subtotal = Column(Numeric(12, 2), nullable=False)
    tax_amount = Column(Numeric(12, 2), default=0.00, nullable=False)
    total_amount = Column(Numeric(12, 2), nullable=False)
    currency = Column(String(3), default="USD", nullable=False)
    
    # Status
    is_paid = Column(Boolean, default=False, nullable=False)
    paid_date = Column(String(30), nullable=True)
    payment_method = Column(String(50), nullable=True)
    
    # Line items
    line_items = Column(JSON, default=list, nullable=False)
    
    # Stripe
    stripe_invoice_id = Column(String(255), nullable=True)
    stripe_payment_intent_id = Column(String(255), nullable=True)
    
    # Relationships
    agency = relationship("Agency", back_populates="invoices")
    
    def __repr__(self):
        return f"<Invoice {self.invoice_number} - ${self.total_amount}>"